--- src/Makefile.in.org	Fri Feb 23 11:34:27 2001
+++ src/Makefile.in	Fri Feb 23 11:33:48 2001
@@ -83,7 +83,7 @@
 PROGRAMS =  $(bin_PROGRAMS)
 
 
-DEFS = @DEFS@ -I. -I$(srcdir) -I..
+DEFS = @DEFS@ -I. -I$(srcdir) -I.. $(RPM_OPT_FLAGS)
 CPPFLAGS = @CPPFLAGS@
 LDFLAGS = @LDFLAGS@
 LIBS = @LIBS@
--- src/IO/Makefile.in.org	Wed Dec 27 23:37:00 2000
+++ src/IO/Makefile.in	Fri Feb 23 11:37:33 2001
@@ -79,7 +79,7 @@
 LIBRARIES =  $(noinst_LIBRARIES)
 
 
-DEFS = @DEFS@ -I. -I$(srcdir) -I../..
+DEFS = @DEFS@ -I. -I$(srcdir) -I../.. $(RPM_OPT_FLAGS)
 CPPFLAGS = @CPPFLAGS@
 LDFLAGS = @LDFLAGS@
 LIBS = @LIBS@
