diff -burNp dillo-2.0/configure.in dillo-2.0-dud/configure.in
--- dillo-2.0/configure.in	2008-10-10 22:37:17.000000000 +0200
+++ dillo-2.0-dud/configure.in	2009-03-24 10:43:37.021819865 +0100
@@ -489,6 +489,11 @@ AC_SUBST(LIBFLTK_CFLAGS)
 AC_SUBST(LIBFLTK_LIBS)
 AC_SUBST(LIBICONV_LIBS)
 AC_SUBST(datadir)
+AC_SUBST(src)
+AC_SUBST(doc)
+AC_SUBST(bin)
+AC_SUBST(util)
+AC_SUBST(lib)
 
 AC_OUTPUT(Makefile dlib/Makefile dpip/Makefile dpid/Makefile dpi/Makefile doc/Makefile dw/Makefile lout/Makefile src/Makefile src/IO/Makefile test/Makefile)
 
